service/Microsoft.Azure.Devices/ServiceClient.cs

```diff
@@ -44,6 +45,26 @@ public static ServiceClient CreateFromConnectionString(string connectionString)
             return CreateFromConnectionString(connectionString, TransportType.Amqp);
         }
 
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Releases unmanaged and - optionally - managed resources.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if ( disposing )
+            {
+                // Call CloseAsync() synchronously avoiding AggregateException.
+                this.CloseAsync().GetAwaiter().GetResult();
+            }
+        }
```

Comment:
In general, that is a valid concern. However, in this case, ServiceClient has an internal parameterless ctor preventing mock objects. 

The real issue is that I believe that dispose should not attempt a graceful close mechanism like CloseAsync(). It should be ungraceful and run to completion without throwing any exceptions. Calling a CloseAsync method which is overridden by derived classes does not allow this.

---
In reply to: [103446925](https://github.com/Azure/azure-iot-sdk-csharp/pull/57#discussion_r103446925) [](ancestors = 103446925)